Frameworks/Foundation/NSURL.mm

```diff
@@ -890,4 +908,63 @@ - (BOOL)getPromisedItemResourceValue:(id _Nullable*)value forKey:(NSString*)key
     return StubReturn();
 }
 
+/**
+ @Status Interoperable
+*/
++ (instancetype)URLWithStorageFile:(RTObject<WSIStorageFile>*)storageFile {
+    return [[[self alloc] initWithStorageFile:storageFile] autorelease];
+}
+
+/**
+ @Status Interoperable
+*/
+- (instancetype)initWithStorageFile:(RTObject<WSIStorageFile>*)storageFile {
+    ComPtr<IInspectable> comPtr = [storageFile comObj];
+    ComPtr<IStorageItem> comStorageItem;
+    RETURN_NULL_IF_FAILED(comPtr.As(&comStorageItem));
+
+    // Now cache the storage item so that it can later be retreived.
+    ComPtr<IStorageApplicationPermissionsStatics> storageApplicationPermissionsStatics;
+    RETURN_NULL_IF_FAILED(
+        GetActivationFactory(Wrappers::HStringReference(RuntimeClass_Windows_Storage_AccessCache_StorageApplicationPermissions).Get(),
+                             &storageApplicationPermissionsStatics));
+
+    ComPtr<IStorageItemAccessList> accessList;
+    RETURN_NULL_IF_FAILED(storageApplicationPermissionsStatics->get_FutureAccessList(&accessList));
+
+    // See how close we are to the cap.
+    uint32_t maxItems = 0;
```

Comment:
it would help to move this to a private function, so the init method isn't doing too many things..